lib/common/socket.c

```diff
@@ -1573,56 +1573,128 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
 #include "h2o-probes.h"
 #include <sys/stat.h>
 
-static __thread int tracing_map_fd = -1;
-static __thread uint64_t tracing_map_last_attempt = 0;
+static int ebpf_map_create(uint32_t map_type, uint32_t key_size, uint32_t value_size, uint32_t max_entries, const char *map_name)
+{
+    union bpf_attr attr = {
+        .map_type = map_type,
+        .key_size = key_size,
+        .value_size = value_size,
+        .max_entries = max_entries,
+    };
+    strncpy(attr.map_name, map_name, sizeof(attr.map_name));
+    return syscall(__NR_bpf, BPF_MAP_CREATE, &attr, sizeof(attr));
+}
+
+static int ebpf_obj_pin(int bpf_fd, const char *pathname)
+{
+    union bpf_attr attr = {
+        .bpf_fd = (uint32_t)bpf_fd,
+        .pathname = (uint64_t)pathname,
+    };
+    return syscall(__NR_bpf, BPF_OBJ_PIN, &attr, sizeof(attr));
+}
+
+static int ebpf_map_lookup(int fd, const void *key, void *value)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+        .value = (uint64_t)value,
+    };
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+}
 
-static void open_tracing_map(h2o_loop_t *loop)
+static int ebpf_map_delete(int fd, const void *key)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+    };
+    return syscall(__NR_bpf, BPF_MAP_DELETE_ELEM, &attr, sizeof(attr));
+}
+
+static int return_map_fd = -1; // for h2o_return
+
+int h2o_socket_ebpf_setup(void)
+{
+    int success = 0;
+    if (getuid() != 0) {
+        h2o_error_printf("failed to set up eBPF maps because bpf(2) requires root privileges\n");
+        goto Exit;
+    }
+
+    /* It creates a pinned BPF object file, and h2o cannot unlink the file because h2o drops root privileges after this function
+     * unless its configuration is set to keep the privileges. */
+    int fd =
+        ebpf_map_create(BPF_MAP_TYPE_LRU_HASH, sizeof(pid_t), sizeof(uint64_t), H2O_EBPF_RETURN_MAP_SIZE, H2O_EBPF_RETURN_MAP_NAME);
+    if (fd < 0) {
+        if (errno == EPERM) {
+            h2o_error_printf("BPF_MAP_CREATE failed with EPERM, "
+                             "maybe because RLIMIT_MEMLOCK is too small.\n");
+        } else {
+            h2o_perror("BPF_MAP_CREATE failed");
+        }
+        goto Exit;
+    }
+
+    if (ebpf_obj_pin(fd, H2O_EBPF_RETURN_MAP_PATH) != 0) {
+        if (errno == EEXIST) {
+            success = 1;
```

Comment:
Ugh...in fact, I misunderstood it, the case where `H2O_EBPF_RETURN_MAP_PATH` already exists. Even the test was wrong. 

If the map already exists, we have to open the file with `BPF_OBJ_GET`, making sure the map's type and name are correct. Fixed dc20115ed

Reply:
FWIW In case the map file already exists, h2o checks its critical attributes, namely type, key size, value size (626a3f448). This check is not to detect attacks but for the changes in the future.  When we changed these attributes in the future, we have to remove the pinned map file on deployment.